File: sdk/src/circuits/olympus.ts
// Olympus - The Pantheon (Domain Registration Circuit)
// Register and manage domains for your Shadow sites

import axios from "axios"

const BACKEND_URL = process.env.SHADOW_BACKEND_URL || "http://localhost:8080"

export interface DomainRegistration {
  domain: string
  programAddress: string
  ownerPubkey: string
}

export interface Domain {
  domain: string
  owner_pubkey: string
  program_address: string
  verified: boolean
  created_at: string
  updated_at: string
}

/**
 * Register a domain to a contract/program address
 * Example: registerDomain("whatsapp.shadow", "WD3EtZGu8Gvhji4GHdLjGBfSc8cGSpyRJ612ANhpump", wallet)
 */
export async function registerDomain(
  domain: string,
  programAddress: string,
  ownerPubkey: string,
  authHeader?: string
): Promise<Domain> {
  const response = await axios.post(
    `${BACKEND_URL}/api/domains`,
    {
      domain,
      program_address: programAddress,
      owner_pubkey: ownerPubkey,
    },
    {
      headers: authHeader ? { "X-Shadow-Auth": authHeader } : {},
    }
  )

  return response.data
}

/**
 * Get domain information by domain name
 */
export async function getDomain(domain: string): Promise<Domain> {
  const response = await axios.get(`${BACKEND_URL}/api/domains/${encodeURIComponent(domain)}`)
  return response.data
}

/**
 * Get domain by program/contract address
 */
export async function getDomainByProgram(programAddress: string): Promise<Domain | null> {
  const response = await axios.get(
    `${BACKEND_URL}/api/domains/search?q=${encodeURIComponent(programAddress)}&limit=1`
  )
  const domains = response.data
  return domains.length > 0 ? domains[0] : null
}

/**
 * Update domain's program address
 */
export async function updateDomain(
  domain: string,
  programAddress: string,
  authHeader: string
): Promise<void> {
  await axios.put(
    `${BACKEND_URL}/api/domains/${encodeURIComponent(domain)}`,
    {
      program_address: programAddress,
    },
    {
      headers: { "X-Shadow-Auth": authHeader },
    }
  )
}

/**
 * Verify domain ownership (after on-chain verification)
 */
export async function verifyDomain(domain: string, authHeader: string): Promise<void> {
  await axios.post(
    `${BACKEND_URL}/api/domains/${encodeURIComponent(domain)}/verify`,
    {},
    {
      headers: { "X-Shadow-Auth": authHeader },
    }
  )
}

/**
 * List all domains owned by a wallet
 */
export async function listOwnerDomains(ownerPubkey: string): Promise<Domain[]> {
  const response = await axios.get(
    `${BACKEND_URL}/api/domains/owner/${encodeURIComponent(ownerPubkey)}`
  )
  return response.data
}

/**
 * Search domains
 */
export async function searchDomains(query: string, limit: number = 10): Promise<Domain[]> {
  const response = await axios.get(
    `${BACKEND_URL}/api/domains/search?q=${encodeURIComponent(query)}&limit=${limit}`
  )
  return response.data
}

File: backend/src/olympus.rs
// Olympus - The Pantheon of Gods (CA Domain System)
// Handles domain registration, verification, and management for Shadow sites

use mongodb::{Collection, Database};
use mongodb::bson::doc;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Domain {
    #[serde(rename = "_id")]
    pub domain: String,                    // e.g., "example.shadow"
    pub owner_pubkey: String,              // Wallet that owns the domain
    pub program_address: String,           // Solana program address it points to
    pub verified: bool,                    // Whether domain is verified on-chain
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub expires_at: Option<DateTime<Utc>>, // Optional expiration
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DomainRegistration {
    pub domain: String,
    pub program_address: String,
    pub owner_pubkey: String,
}

pub struct OlympusCA {
    db: Database,
}

impl OlympusCA {
    pub fn new(db: Database) -> Self {
        Self { db }
    }

    fn get_domains_collection(&self) -> Collection<Domain> {
        self.db.collection::<Domain>("domains")
    }

    /// Register a new domain (Pantheon entry)
    /// Maps a domain to a Solana program/contract address
    pub async fn register_domain(
        &self,
        domain: &str,
        owner_pubkey: &str,
        program_address: &str,
    ) -> Result<(), String> {
        // Domain validation is done by Apollo, just check it's not empty
        if domain.is_empty() {
            return Err("Domain cannot be empty".to_string());
        }

        let collection = self.get_domains_collection();
        let now = Utc::now();
        let bson_now = mongodb::bson::DateTime::from_millis(now.timestamp_millis());

        let filter = doc! { "_id": domain };
        let update = doc! {
            "$set": {
                "owner_pubkey": owner_pubkey,
                "program_address": program_address,
                "verified": false,
                "updated_at": bson_now
            },
            "$setOnInsert": {
                "created_at": bson_now
            }
        };

        let options = mongodb::options::UpdateOptions::builder()
            .upsert(true)
            .build();

        collection.update_one(filter, update, options).await
            .map_err(|e| format!("Database error: {}", e))?;

        Ok(())
    }

    /// Get domain by name
    pub async fn get_domain(&self, domain: &str) -> Result<Option<Domain>, String> {
        let collection = self.get_domains_collection();
        let filter = doc! { "_id": domain };
        
        collection.find_one(filter, None).await
            .map_err(|e| format!("Database error: {}", e))
    }

    /// Get domain by program address
    pub async fn get_domain_by_program(&self, program_address: &str) -> Result<Option<Domain>, String> {
        let collection = self.get_domains_collection();
        let filter = doc! { "program_address": program_address };
        
        collection.find_one(filter, None).await
            .map_err(|e| format!("Database error: {}", e))
    }

    /// Verify domain ownership (mark as verified after on-chain verification)
    pub async fn verify_domain(&self, domain: &str) -> Result<(), String> {
        let collection = self.get_domains_collection();
        let now = Utc::now();
        let bson_now = mongodb::bson::DateTime::from_millis(now.timestamp_millis());

        let filter = doc! { "_id": domain };
        let update = doc! {
            "$set": {
                "verified": true,
                "updated_at": bson_now
            }
        };

        collection.update_one(filter, update, None).await
            .map_err(|e| format!("Database error: {}", e))?;

        Ok(())
    }

    /// Transfer domain ownership
    pub async fn transfer_domain(
        &self,
        domain: &str,
        new_owner: &str,
    ) -> Result<(), String> {
        let collection = self.get_domains_collection();
        let now = Utc::now();
        let bson_now = mongodb::bson::DateTime::from_millis(now.timestamp_millis());

        let filter = doc! { "_id": domain };
        let update = doc! {
            "$set": {
                "owner_pubkey": new_owner,
                "verified": false, // Require re-verification after transfer
                "updated_at": bson_now
            }
        };

        collection.update_one(filter, update, None).await
            .map_err(|e| format!("Database error: {}", e))?;

        Ok(())
    }

    /// List domains owned by a wallet
    pub async fn list_owner_domains(&self, owner_pubkey: &str) -> Result<Vec<Domain>, String> {
        let collection = self.get_domains_collection();
        let filter = doc! { "owner_pubkey": owner_pubkey };
        
        let mut cursor = collection.find(filter, None).await
            .map_err(|e| format!("Database error: {}", e))?;

        let mut domains = Vec::new();
        use futures_util::TryStreamExt;
        while let Some(domain) = cursor.try_next().await
            .map_err(|e| format!("Database error: {}", e))? {
            domains.push(domain);
        }

        Ok(domains)
    }

    /// Search domains
    pub async fn search_domains(&self, query: &str, limit: i64) -> Result<Vec<Domain>, String> {
        let collection = self.get_domains_collection();
        let filter = doc! {
            "$or": [
                { "_id": { "$regex": query, "$options": "i" } },
                { "program_address": { "$regex": query, "$options": "i" } }
            ],
            "verified": true
        };

        let options = mongodb::options::FindOptions::builder()
            .limit(limit)
            .sort(doc! { "created_at": -1 })
            .build();

        let mut cursor = collection.find(filter, options).await
            .map_err(|e| format!("Database error: {}", e))?;

        let mut domains = Vec::new();
        use futures_util::TryStreamExt;
        while let Some(domain) = cursor.try_next().await
            .map_err(|e| format!("Database error: {}", e))? {
            domains.push(domain);
        }

        Ok(domains)
    }
}

